subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyAdder.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.dsl.dependencies;
+
+import org.gradle.api.Action;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.ExternalModuleDependency;
+import org.gradle.api.artifacts.ExternalModuleDependencyBundle;
+import org.gradle.api.artifacts.FileCollectionDependency;
+import org.gradle.api.artifacts.MinimalExternalModuleDependency;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.artifacts.dsl.DependencyAdder;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderConvertible;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class DefaultDependencyAdder implements DependencyAdder {
+    private final DependencyFactoryInternal dependencyFactory;
+    private final Configuration configuration;
+
+    @Inject
+    public DefaultDependencyAdder(DependencyFactoryInternal dependencyFactory, Configuration configuration) {
+        this.dependencyFactory = dependencyFactory;
+        this.configuration = configuration;
+    }
+
+    @SuppressWarnings("unchecked")
+    private <D extends Dependency> D finalizeDependency(D dependency) {
+        if (dependency instanceof MinimalExternalModuleDependency) {
+            // This will break if D == MinimalExternalModuleDependency
```

Comment:
What does this mean?  isn't this checking for when D == MinimalExternalModuleDependency?

Reply:
With the recent commits this is no longer true, but basically you could have `D = ExternalModuleDependency` but `dependency instanceof MEMD` and that would work; but `D = MinimalExternalModuleDependency` would not have because the configuration actions above would target the wrong type and get a `ClassCastException`.

I've removed this comment now.